### DISCUSSION TITLE
Shouldn't `safetensors` allow the storage of e.g. a `flax.frozen_dict.FrozenDict`?

### DISCUSSION THREAD

**Discussion by alvarobartt** [huggingface/safetensors#138] [Q&A] [answered]
AFAIU the main idea behind `safetensors` is to define a more general and safe/efficient format to store tensors. But one question comes to my mind, shouldn't it be useful not to just store plain tensors but also more complex structures containing tensors such as `flax.frozen_dict.FrozenDict`?

Just asking because it's unclear to me in which direction is the library going or whether a more general serialization format is the end goal or not.

Thanks! :hugs:

**Comment by Narsil** [comment, marked as answer]
Hey thanks for your question.

Currently there's no plan to support anything more than tensors.

The initiation for this lib was simply sparked to try and avoid using `pickle`. As the README suggests other libraries where considered but all had issues that prevented them from being used (`H5` was the biggest contendant, but since TF was deprecating its use and it doesn't support `bfloat16` which is kind of important for models like `bloom` we didn´t  go that way).

The goal of this lib is to be trivial  (so you can actually load `safetensors` files without even using this lib : https://gist.github.com/Narsil/3edeec2669a5e94e4707aa0f901d2282

And since we're creating something new (https://xkcd.com/927/) we might as well add some features which were missing from everything else:

- Slice support (you can load only part of a tensor and doing so efficiently, this is important for very LLMs like bloom where you need to load only parts of the models on each GPU).
- Slice support for writing (this is for training LLMs and is in the works)
- Faster loading for both GPU and CPU. Since we can have a very flat layout we can optimize quite a bit the loading of the tensors which wouldn't be the case with more complex data structures.
- In addition to arbitrary code execution (ACE) prevention, I decided to go for DOS (denial of service) prevention, so this library should not DOS you even with maliciously crafted files. (The file will take at most its size in memory)

Does that answer your question ?

**Reply by alvarobartt** [reply]
Sure @Narsil thanks for the detailed answer! I was exploring `safetensors` to integrate it with something I'm building on top of `flax`, and I wanted to see whether this was compatible with `flax.serialization`.

**Reply by Narsil** [reply]
`safetensors.flax.load_file` should work out of the box (if the file contains directly tensors)

**Reply by alvarobartt** [reply]
Yes, it does the work, but e.g. if you define a `flax.linen.nn.Module` with a single `Dense` layer, you'll have both the weights and the biases, so you end up having a dict of tensors formatted as `{"Dense0": {"weights": jnp.DeviceArray(...), "bias": jnp.DeviceArray(...)}` so you'll be able to save a single layer when selecting `Dense0`, but not to save everything at once. So in case, you want to use `safetensors` to serialize the whole params/weights dictionary, that's not an option right now.

**Reply by alvarobartt** [reply]
Also as `safetensors` is intended just for tensors and not for more complex structures, the conversion at `_jnp2np` is not recursive, which means that will just convert to `np.ndarray` the first-level structures, which in the case mentioned above are not `jnp.DeviceArray` tensors, but another dict with more than one tensor inside.

**Reply by alvarobartt** [reply]
In any case, if you do a recursive transformation then the endianness won't match as it won't be little endian (<) but not applicable (|), at https://numpy.org/doc/stable/reference/generated/numpy.dtype.byteorder.html

**Reply by Narsil** [reply]
Sorry I don't really understand.

Denesting/Nesting the structures is rather trivial but has to be done by users yes (just use `.` as separator since you cannot use `.` within Python variable, this is how it's done in `transformers` for instance).
Something like that for instance: https://gist.github.com/Narsil/d5b0d747e5c8c299eb6d82709e480e3d

(This library's goal is to propely load and save tensors, not deal with other formats like msgpack and the format flax uses)

> In any case, if you do a recursive transformation then the endianness won't match as it won't be little endian (<) but not applicable (|), at https://numpy.org/doc/stable/reference/generated/numpy.dtype.byteorder.html

Endianness is enforced to be little-endian: https://github.com/huggingface/safetensors/blob/main/bindings/python/py_src/safetensors/numpy.py#L37

**Reply by alvarobartt** [reply]
Oh ok, so ideally denesting/nesting structures whether it's a `pytree` or `FrozenDict` is supposed to be done by users, that's what was unclear to me, I already did something about that for `FrozenDict` yesterday, thanks again for the detailed answer, it's more clear now!

**Reply by alvarobartt** [reply]
Thanks for this https://gist.github.com/Narsil/d5b0d747e5c8c299eb6d82709e480e3d! I had something similar that I was using for simple structures, didn't test it extensively anyway, I'll probably build something simple around that in case any user is looking for similar answers. I was also having a look at what you did at https://huggingface.co/spaces/safetensors/convert/blob/main/convert.py, in case I develop something I'll just tag it here so that anyone can use it. Great work and thanks again! :hugs:

**Reply by Narsil** [reply]
Thanks a lot ! 

`convert.py` stemmed from a need to convert pytorch checkpoints, but it's starting to become quite big, I think having 2 files `convert_pt.py` and `convert_flax.py` might be better for readability and usability (dependencies).

**Reply by alvarobartt** [reply]
I'm already working on something, I think that this week I'll be able to release the first dev version of the package that I created to use `safetensors` to both save and load `flax.core.frozen_dict.FrozenDict`. I'll probably post it here in the replies if it's useful for anyone! Do you also want me to collaborate in that HF Space with a `convert_flax.py` file or something? Thanks for your work, really enjoying `safetensors` usage and efficiency so far!

**Reply by Narsil** [reply]
Sure let us know when you're done for the script and/or the space !

**Reply by alvarobartt** [reply]
Thanks for your help @Narsil, project is already available at https://github.com/alvarobartt/safejax, install it as `pip install safejax` 👍🏻
